TextIterator & StyledMarkupAccumulator should consider skipped content as invisible

nt1m opened the pull request:
#### 8d60cd8712f39a96e7bd61fd006af22fd76a5c2a
<pre>
TextIterator &amp; StyledMarkupAccumulator should consider skipped content as invisible
<a href="https://bugs.webkit.org/show_bug.cgi?id=283344">https://bugs.webkit.org/show_bug.cgi?id=283344</a>
<a href="https://rdar.apple.com/140172890">rdar://140172890</a>

Reviewed by Ryosuke Niwa.

When content is skipped as a result of content-visibility: hidden/auto, it is not visible to the user.

Make sure we properly skip the content from selection or copy &amp; paste to avoid privacy leaks.

innerText should also avoid skipped content as per spec: <a href="https://drafts.csswg.org/css-contain/#cv-notes">https://drafts.csswg.org/css-contain/#cv-notes</a>

* LayoutTests/editing/pasteboard/data-transfer-get-data-on-paste-skipped-content-expected.txt: Added.
* LayoutTests/editing/pasteboard/data-transfer-get-data-on-paste-skipped-content.html: Added.
* LayoutTests/editing/pasteboard/data-transfer-set-data-sanitizes-html-when-copying-expected.txt:
* LayoutTests/editing/pasteboard/data-transfer-set-data-sanitizes-html-when-copying-in-null-origin-expected.txt:
* LayoutTests/editing/pasteboard/data-transfer-set-data-sanitizes-html-when-copying-in-null-origin.html:
* LayoutTests/editing/pasteboard/data-transfer-set-data-sanitizes-html-when-copying.html:
* LayoutTests/editing/selection/selection-toString-skipped-content-expected.txt: Added.
* LayoutTests/editing/selection/selection-toString-skipped-content.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-035-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-044-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-053-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-relevancy-updates-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-053-expected.txt: Added.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-053-expected.txt: Added.
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::isRendererVisible):
* Source/WebCore/editing/markup.cpp:
(WebCore::StyledMarkupAccumulator::traverseNodesForSerialization):

Canonical link: <a href="https://commits.webkit.org/286799@main">https://commits.webkit.org/286799@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f472633207529c006f916a80f8f23313d7136fbe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77119 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56154 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30034 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/81677 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28401 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79236 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65302 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4450 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/81677 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/18493 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80186 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/50390 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/66198 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/81677 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/47792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/23696 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26724 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/68907 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/24024 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/83104 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4499 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/83104 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4655 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/66171 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/83104 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/11948 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/10035 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11935 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4445 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7261 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4465 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7900 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6224 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/editing/markup.cpp:
```diff
@@ -774,6 +774,9 @@ RefPtr<Node> StyledMarkupAccumulator::traverseNodesForSerialization(Node& startN
         if (!node.renderer() && !isDisplayContents && !enclosingElementWithTag(firstPositionInOrBeforeNode(&node), selectTag))
             return false;
 
+        if (node.renderer() && node.renderer()->isSkippedContent())
```
We need to add a test for this.

rniwa replied:
We can combine this with the previous if so that we don't keep checking node.renderer():
```cpp
if (node.renderer()) {
    if (node.renderer()->isSkippedContent())
        return false;
} else if (!isDisplayContents && !enclosingElementWithTag(firstPositionInOrBeforeNode(&node), selectTag))
    return true;
```

Or maybe even just:
```cpp
if ((node.renderer() && node.renderer()->isSkippedContent()) || (!node.renderer() && !isDisplayContents !enclosingElementWithTag(firstPositionInOrBeforeNode(&node), selectTag)))
    return false;
```

rniwa commented on LayoutTests/editing/pasteboard/data-transfer-get-data-on-paste-skipped-content.html:
```diff
@@ -0,0 +1,47 @@
+{
+    shouldBeEqualToString('JSON.stringify(event.clipboardData.types)', '["text/plain"]');
+    shouldBeEqualToString('event.clipboardData.getData("text/plain")', 'hello, world\nWebKit');
+    shouldBeEqualToString('event.clipboardData.getData("text/html")', '');
```
Why is text/html empty? That doesn't sound right at all.

nt1m replied:
This is a literal copy paste of `LayoutTests/editing/pasteboard/data-transfer-get-data-on-paste-as-plain-text.html`. I'm not sure what's going on there to be honest. Is it maybe related to `internals.settings.setCustomPasteboardDataEnabled(true);` ?

nt1m committed:
TextIterator & StyledMarkupAccumulator should consider skipped content as invisible
https://bugs.webkit.org/show_bug.cgi?id=283344
rdar://140172890

Reviewed by Ryosuke Niwa.

When content is skipped as a result of content-visibility: hidden/auto, it is not visible to the user.

Make sure we properly skip the content from selection or copy & paste to avoid privacy leaks.

innerText should also avoid skipped content as per spec: https://drafts.csswg.org/css-contain/#cv-notes

* LayoutTests/editing/pasteboard/data-transfer-get-data-on-paste-skipped-content-expected.txt: Added.
* LayoutTests/editing/pasteboard/data-transfer-get-data-on-paste-skipped-content.html: Added.
* LayoutTests/editing/pasteboard/data-transfer-set-data-sanitizes-html-when-copying-expected.txt:
* LayoutTests/editing/pasteboard/data-transfer-set-data-sanitizes-html-when-copying-in-null-origin-expected.txt:
* LayoutTests/editing/pasteboard/data-transfer-set-data-sanitizes-html-when-copying-in-null-origin.html:
* LayoutTests/editing/pasteboard/data-transfer-set-data-sanitizes-html-when-copying.html:
* LayoutTests/editing/selection/selection-toString-skipped-content-expected.txt: Added.
* LayoutTests/editing/selection/selection-toString-skipped-content.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-035-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-044-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-053-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-relevancy-updates-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-053-expected.txt: Added.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-053-expected.txt: Added.
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::isRendererVisible):
* Source/WebCore/editing/markup.cpp:
(WebCore::StyledMarkupAccumulator::traverseNodesForSerialization):

Canonical link: https://commits.webkit.org/286799@main
